cql3/restrictions/analyzed_where.hh

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#pragma once
+
+#include "index_table_query_restrictions.hh"
+#include "single_table_query_restrictions.hh"
+#include "statement_restrictions.hh"
+
+namespace cql3 {
+namespace restrictions {
+
+// This class analyzes the WHERE clause and generates a query plan.
+class analyzed_where_clause {
+    // The query is a single_table_query if possible without filtering, otherwise it might be an indexed query.
+    std::optional<std::variant<single_table_query_restrictions, index_table_query_restrictions>> query_restrictions;
+
+    // For now we also keep the old implementation using statement_restrictions.
+    // Its functionality will be gradually replaced by single_table_query_restrictions and
+    // index_table_query_restrictions
+    ::shared_ptr<cql3::restrictions::statement_restrictions> restrictions;
+
+   public:
+    analyzed_where_clause(data_dictionary::database db,
+                          schema_ptr schema,
+                          statements::statement_type type,
+                          const expr::expression& where_clause,
+                          prepare_context& ctx,
+                          bool selects_only_static_columns,
+                          bool for_view = false,
+                          bool allow_filtering = false);
+
+    // Creates analyzed_where_clause for an empty WHERE.
+    analyzed_where_clause(schema_ptr schema, bool allow_filtering);
```

Comment:
I feel that this should be a function that returns an analyzed_where_clause, since this is doing the bulk of the work. But it's just cosmetic.

Reply:
My thinking was `I want a function that is used to create an instance of this class`, so a constructor was a natural choice. I like classes with constructors and methods more than freestanding functions, because all functionality is easily findable in one place. Are constructors that do a lot of work a bad thing?
OTOH lots of expression code is freestanding functions, so maybe this should be one as well for consistency.
I don't have any strong opinion about it.

Reply:
> My thinking was `I want a function that is used to create an instance of this class`, so a constructor was a natural choice. I like classes with constructors and methods more than freestanding functions, because all functionality is easily findable in one place. Are constructors that do a lot of work a bad thing? OTOH lots of expression code is freestanding functions, so maybe this should be one as well for consistency. I don't have any strong opinion about it.

I can't say objectively that constructors that do a lot of work are bad. Subjectively I feel that a function is a verb that does something, while a constructor just assembles a bigger thing from some small thing. It's especially true if all the class contains is the constructor and data members.